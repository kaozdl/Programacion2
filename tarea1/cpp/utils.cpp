/* 4580203 */
#include "../include/utils.h"

#include <stdio.h>

void imprimir_comp(const comp_t comp) {
  switch (comp) {
  case menor: {
    printf("<");
    /* En general, debe usarse break para evitar que siga procesando los otros
       casos. */
    break;
  }
  case igual: {
    printf("=");
    break;
  }
  case mayor: {
    printf(">");
    break;
  }
  }
} // imprimir_comp
