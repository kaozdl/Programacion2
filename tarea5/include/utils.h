/*
  Módulo de definición de `utils`.

  Se define un tipo genérico, T.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#ifndef _UTILS_H
#define _UTILS_H

enum comp_t { menor, igual, mayor }; // comparación

/*
 Imprime `<`, '=' o `>`  en la salida estándar si `comp` es `menor`, `igual`
 o `mayor`respectivamente.
 */
void imprimir_comp(const comp_t comp);

typedef unsigned int nat; // naturales

#endif
