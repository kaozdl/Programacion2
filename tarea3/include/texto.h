/*
  Módulo de definición de `texto`.

  'texto_t' es un puntero a una estructura (opaca, que debe definirse en
  texto.cpp) cuyo contenido es un texto (arreglo de elementos de tipo char).

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#ifndef _TEXTO_H
#define _TEXTO_H

#include "utils.h"

/* representación de `texto`, definida en texto.cpp */
struct rep_texto;

/* El tipo `texto_t` es un puntero a elementos de tipo `rep_texto`. */
typedef rep_texto *texto_t;

/* **************  NUEVO  ************** */

/*
  Devuelve el resultado de convertir el contenido de `t` en comp_t.
  Precondición: el contenido de `t` es `<`, `=` o `>`.
 */
comp_t texto_a_comp(const texto_t txt);

/*
  Devuelve un `texto` cuyo contenido es vacío (de longitud 0).
 */
texto_t texto_vacio();

/*
  Devuelve un `texto` cuyo contenido es `simbolo`.
 */
texto_t char_a_texto(const char simbolo);

/*
  Devuelve un `texto` cuyo contenido es una copia de `cadena`.
  El elemento devuelto no comparte memoria con `cadena`.
  Precondición: `cadena != NULL`.
 */
texto_t cadena_a_texto(const char *cadena);

/*
  Devuelve un `texto` leído desde la entrada estándar luego de ignorar los
  espacios en blanco iniciales.
  Lee hasta encontrar un espacio en blanco.
  Precondición: la palabra tiene a lo sumo `max` caracteres.
 */
texto_t leer_palabra(const nat max);

/*
  Devuelve un `texto` leído desde la entrada estándar.
  Lee hasta encontrar el carácter de `fin de línea` o de `fin de archivo`.
  Ese carácter es leído pero no se incluye en el `texto` devuelto.
  Precondición: la línea tiene a lo sumo `max` caracteres, incluido el 'fin de
  línea' o 'fin de archivo'.
 */
texto_t leer_resto_linea(const nat max);

/*
  Escribe en la salida estándar el contenido de `t`.
 */
void escribir_texto(const texto_t t);

/*
  Inserta en la salida estándar `nueva línea` (inserta una línea y mueve el
  cursor a la primera columna de la línea insertada).
 */
void escribir_nueva_linea();

/*
  Devuelve un `texto` cuyo contenido es la concatenación de las copias de
  los de `t1` y `t2`.
  El elemento devuelto no comparte memoria ni con `t1` ni con `t2`.
 */
texto_t concatenar_texto(const texto_t t1, const texto_t t2);

/*
  Devuelve un `texto` cuyo contenido es una copia del de `t`.
  El elemento devuelto no comparte memoria con `t`.
 */
texto_t copiar_texto(const texto_t t);

/*
  Devuelve `true` si `t` es vacío o `false` en otro caso.
 */
bool es_texto_vacio(const texto_t t);

/*
  Devuelve la longitud del contenido de `t`.
 */
nat longitud_texto(const texto_t t);

/*
  Devuelve el resultado de la comparación lexicográfica de los contenidos de
  `t1` y `t2`.
 */
comp_t comparar_texto(const texto_t t1, const texto_t t2);

/*
  Devuelve el resultado de convertir el contenido de `t` en un entero.
 */
int texto_a_int(const texto_t t);

/*
 Devuelve el resultado de convertir `i` a texto_t.
 */
texto_t int_a_texto(const int i);

/*
  Libera la memoria asignada a `t`.
 */
void liberar_texto(texto_t &t);

#endif
