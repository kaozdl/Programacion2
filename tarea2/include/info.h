/*
  Módulo de definición de `info`.

  `info_t` es un puntero a una estructura (que debe definirse en
  info.cpp) cuyo contenido es un dato numérico (entero) y un dato de texto.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
 */

#ifndef _INFO_T
#define _INFO_T

#include "utils.h"
#include "texto.h"

struct rep_info; // representación de `info`, definida en info.cpp
typedef rep_info *info_t;

/* Constructora, destructura y selectoras de info_t. */

/*
  Devuelve un info_t tal que su dato numérico es `n` y su dato de texto
  es `t`.
 */
info_t crear_info(const int n, const texto_t t);

/*
  Devuelve una copia de `i`.
  El elemento devuelto no comparte memoria con `i`.
 */
/*
info_t copiar_info(const info_t i);
Se incluirá en la próxima tarea.
*/

/*
  Libera la memoria reservada por `i`.
 */
void liberar_info(info_t &i);

/*
  Devuelve el dato numérico asociado a `i`.
 */
int numero_info(const info_t i);

/*
  Devuelve el dato de texto asociado a `i`.
 */
texto_t texto_info(const info_t i);

/*
  Devuelve `true` si `i` es válida o `false` en otro caso.
  `i` no es válida cuando su valor numérico es INT_MAX.
 */
bool es_valida_info(const info_t i);

/*
  Devuelve un texto_t formado por la concatenacion del dato numérico de `i`,
  el string "," y el dato de texto de `i` encerrados entre parentesis.
  El elemento devuelto no comparte memoria con `i`.
   Ejemplos:
   (4,a)
   (56,jojo)
 */
texto_t info_a_texto(const info_t i);

/*
  Lee texto desde la entrada estándar. Si tiene formato de `info_t` (ver
  info_a_texto) lo devuelve convertido en `info_t`. En otro caso devuelve una
  `info_t` con un  dato numérico = INT_MAX y dato de texto vacío.
  Precondición: hay texto para leer, que no incluye el símbolo ')' y tiene a
  lo sumo `max` caracteres.
 */
info_t leer_info(nat max);

#endif
