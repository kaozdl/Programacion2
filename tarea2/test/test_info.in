(50,test de info_t) .
