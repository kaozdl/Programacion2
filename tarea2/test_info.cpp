/*
  Módulo principal test_info.

  Laboratorio de Programación 2.
  InCo-FIng-UDELAR
*/

#include "include/utils.h"
#include "include/texto.h"
#include "include/info.h"

#include <stdio.h>

#define MAX_TEXTO 100

int main() {
  texto_t texto = cadena_a_texto("-Este dato de texto-");
  info_t info = crear_info(20, texto);
  printf("El dato numerico es %d.\n", numero_info(info));
  fputs("El dato de texto es ", stdout);
  texto_t txt2 = texto_info(info);
  escribir_texto(txt2);
  escribir_nueva_linea();
  fputs("Info a texto es ", stdout);
  txt2 = info_a_texto(info);
  escribir_texto(txt2);
  escribir_nueva_linea();

  liberar_texto(txt2);
  liberar_info(info);

  printf("Ingrese info (con el formato de info_a_texto) seguido de un punto.\n "
         "Ejemplo: (10,hola mundo):\n ");
  info = leer_info(MAX_TEXTO);
  printf("El dato numerico es %d\n", numero_info(info));
  fputs("El dato de texto es ", stdout);
  escribir_texto(texto_info(info));
  escribir_nueva_linea();
  liberar_info(info);

  return 0;
}
