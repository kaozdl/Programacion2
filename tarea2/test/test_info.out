El dato numerico es 20.
El dato de texto es -Este dato de texto-
Info a texto es (20,-Este dato de texto-)
Ingrese info (con el formato de info_a_texto) seguido de un punto.
 Ejemplo: (10,hola mundo):
 El dato numerico es 50
El dato de texto es test de info_t
